pkg/otlp/model/translator/statspayload.go

```diff
@@ -0,0 +1,471 @@
+)
+
+// keyAPMStats specifies the key name of the resource attribute which identifies resource metrics
+// as being an APM Stats Payload. The presence of the key results in them being treated and consumed
+// differently by the Translator.
+const keyAPMStats = "_dd.apm_stats"
+
+// This group of constants specifies the names used to encode APM Stats as Metrics.
+const (
+	statsKeyHostname         = "dd.hostname"
+	statsKeyEnv              = "dd.env"
+	statsKeyVersion          = "dd.version"
+	statsKeyLang             = "dd.lang"
+	statsKeyTracerVersion    = "dd.tracer_version"
+	statsKeyRuntimeID        = "dd.runtime_id"
+	statsKeySequence         = "dd.sequence"
+	statsKeyAgentAggregation = "dd.agent_aggregation"
+	statsKeyService          = "dd.service"
+	statsKeyContainerID      = "dd.container_id"
+	statsKeyTags             = "dd.tags"
+	statsKeySynthetics       = "dd.synthetics"
+	statsKeySpanName         = "dd.name"
+	statsKeySpanResource     = "dd.resource"
+	statsKeyHTTPStatusCode   = "dd.http_status_code"
+	statsKeySpanType         = "dd.type"
+	statsKeySpanDBType       = "dd.db_type"
+)
+
+// StatsPayloadToMetrics converts an APM Stats Payload to a set of OTLP Metrics.
+func (t *Translator) StatsPayloadToMetrics(sp pb.StatsPayload) pmetric.Metrics {
+	mmx := pmetric.NewMetrics()
+	// We ignore Agent{Hostname,Env,Version} and fill those in later. We want those
+	// values to be consistent with the ones that appear on traces and logs. They are
+	// only known in the Datadog exporter or the Datadog Agent OTLP Ingest.
+	var npayloads, nbuckets, ngroups int
+	for _, cp := range sp.Stats {
+		npayloads++
+		rmx := mmx.ResourceMetrics().AppendEmpty()
+		attr := rmx.Resource().Attributes()
+		attr.PutBool(keyAPMStats, true)
+		putStr(attr, statsKeyHostname, cp.Hostname)
+		putStr(attr, statsKeyEnv, cp.Env)
+		putStr(attr, statsKeyVersion, cp.Version)
+		putStr(attr, statsKeyLang, cp.Lang)
+		putStr(attr, statsKeyTracerVersion, cp.TracerVersion)
+		putStr(attr, statsKeyRuntimeID, cp.RuntimeID)
+		putInt(attr, statsKeySequence, int64(cp.Sequence))
+		putStr(attr, statsKeyAgentAggregation, cp.AgentAggregation)
+		putStr(attr, statsKeyService, cp.Service)
+		putStr(attr, statsKeyContainerID, cp.ContainerID)
+		putStr(attr, statsKeyTags, strings.Join(cp.Tags, ","))
+
+		for _, sb := range cp.Stats {
+			nbuckets++
+			smx := rmx.ScopeMetrics().AppendEmpty()
+			for _, cgs := range sb.Stats {
+				ngroups++
+				mxs := smx.Metrics()
+				for name, val := range map[string]uint64{
+					"hits":           cgs.Hits,
+					"errors":         cgs.Errors,
+					"duration":       cgs.Duration,
+					"top_level_hits": cgs.TopLevelHits,
+				} {
+					appendSum(mxs, name, int64(val), sb.Start, sb.Start+sb.Duration, &cgs)
+				}
+				if err := appendSketch(mxs, "ok_summary", cgs.OkSummary, sb.Start, sb.Start+sb.Duration, &cgs); err != nil {
+					t.logger.Error("StatsPayload Translator: error exporting ok_summary", zap.Error(err))
```

Comment:
Instead of prefixing logs with `StatsPayload Translator`, add a new attribute referring to this (you could even create a new logger from `t.logger` that always appends this field using the [`With`](https://pkg.go.dev/go.uber.org/zap#Logger.With) method)

Reply:
I created a child logger with a `component: metrics translator` field and toned down on the prefixes.